### ISSUE TITLE
browse.path not set automatically when using compile_commands.json

### ISSUE THREAD

**Issue by MathiasMagnus** [microsoft/vscode-cpptools#1163]
I have VS Code 1.17.2 with CppTools 0.14.0 and CMakeTools 0.10.2 and there are some problems with the IntelliSense engine lighting up.

Relevant entries from settings.json

```json
// CMake Configuration
  "cmake.cmakePath": "C:/Kellekek/Kitware/CMake/3.9.0/bin/cmake.exe",

// C/C++ Configuration
  "C_Cpp.intelliSenseEngineFallback": "Disabled",
  "C_Cpp.loggingLevel": "Information"
```

Because all my C++ projects are CMake based, I turn off the Tag-based engine, because I want to make sure. IntelliSense errors should always be compilation errors (as it usually is in big brother VS).

In c_cpp_propertires.json under `Win32` I got

```json
"compileCommands": "${workspaceRoot}/build/compile_commands.json"
```

Once I do this, symbols show in the hover tooltips correctly (`std::vector<T, alloc<T>>::cend()` shows properly with correct return type), but trying to navigate there in code does not work. The blue tooltip says `No definition found for 'cend'`.

Am I doing something wrong?

**Comment by MathiasMagnus** [comment]
The [log.txt](https://github.com/Microsoft/vscode-cpptools/files/1410805/log.txt) of the extension _only_ argues about

```
Unable to retrieve file system information for C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/include/*. error = -1
```

which seems fair enough, because the STL headers are not located in that folder, because VS is installed in a non-default location, hence there's nothing to find in the folder where the log entry suggests to find system information.

**Comment by bobbrow** [comment]
code browsing (navigation) is still powered by the tag parser engine, so you still need to set `"browse.path"` for the time being.

**Comment by MathiasMagnus** [comment]
Thanks @bobbrow for the heads up. I am preparing slides for a seminar series at my institute on Nov. 20-21 where I wanted to demo the new awesome in VS Code, how I finally caught up to big brother VS in terms of C++ usability in the context of CMake. It seems code navigation did not make it into this release. Could I hope to incorporate this feature if I made my slides on the 19th? :)

_Really, the less json I have to show to people, the better the impression is. I am surprised that only the hotkey menu received a GUI, but not the ordinary settings.json._

**Comment by bobbrow** [comment]
Unfortunately, this will not be ready in November.  When opening a new folder without a c_cpp_properties.json file in it, we should (hopefully) be picking a reasonable set of defaults for you (system includes + workspaceRoot).

**Comment by Flamefire** [comment]
I have the same (or very similar) problem (cpptools + CMake tools):
- Open a CMake project & Configure with CMake tools
- Set `compileCommands` to the `compile_commands.json`
- Add `...foo/include` to `browse.path` (Shows up as `Folder: ...foo/include/ will be indexed`)

Results:
- Navigate to include file included with `#include <foo.hpp>` -> Works (after `browse.path`)
- Hover over function -> Works (function declaration shown)
- Click `Go to declaration` -> Not working (`No definition found`)

**Comment by bobbrow** [comment]
@Flamefire, can you provide a small example where this is the case?  Note: that if you're trying to do goto def on a symbol scoped by a namespace alias (and maybe also a typedef), this doesn't work currently.  [0.17.7-insiders](https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.17.7-insiders) has a fix for the namespace alias issue.

For example:
```cpp
namespace N {
    class Foo {};
}
namespace Alias = N;
Alias::Foo f;  // GotoDef/Decl on 'Foo' doesn't work in 0.17.6
```

Many other GotoDef/Decl cases should be fixed with #255 which is one of the next big items on our backlog.

**Comment by Flamefire** [comment]
Not really, the code I'm using is not open source (yet) and I can't reduce this. I'll try to describe what I can:

Work dir: ~/git/project/subdir/
Build dir: ~/build/project
Compile command (from compile_commands.json): `"/usr/bin/g++-6  -DBOOST_VARIANT_DO_NOT_USE_VARIADIC_TEMPLATES -I~/git/project/subdir/component/.. -isystem ~/foo/include   -Wall -Wextra -Werror -O3 -DNDEBUG   -std=gnu++11 -o CMakeFiles/project.dir/component.cpp.o -c ~/git/project/subdir/component/component.cpp"` (home directory replaced manually for github)

In ~/foo/include I have (among others) 2 files:
- foo/class1.hpp
- foo/class2.hpp (includes class1)

All headers use `#pragma once`, class1&2 are in a namespace `foo::bar` (no alias).

In `component.cpp` I have multiple includes like `#include <foo/class2.hpp>`. I can `goto` class1, although I did not include its header. But I cannot `goto` class2.

I tried adding the include to class1 just above the class2 include and noticed something strange: CTRL+click on class1 include opens the class2 header. So it seams that some internal database is outdated. Although the goto behavior is unchanged.

Anything else I could do to gather more input?

**Comment by bobbrow** [comment]
Things I can rule out:
* compile_commands.json does not currently affect goto definition.
* `#include` does not affect goto definition (unless the header lives outside of `${workspaceFolder}`). Symbols are found based on the value of `browse.path` (or `includePath` if `browse.path` is not present).

In the case where you added the new `#include` line, did you save the file before CTL+click?  The DB is currently not updated until you save your changes so that might explain what happened.

If I had to make a guess, the problem is that class2 isn't in the database because the tag parser failed somewhere while processing the header.  You are compiling with `-std=gnu++11`, so maybe there's a GCC extension we don't handle...  Can you check by opening class2.hpp and putting the text cursor somewhere inside the definition of class2?  The status bar should show which class/function you're in.  If it says something wrong, like "(Global Scope)" or some other symbol, then that would confirm my suspicion.

**Comment by Flamefire** [comment]
The headers of class1/2 are outside the workspaceFolder if this matters.

Yes I did save. I also get the out "Checking for syntax errors" message for that file, but only, when I switch the editor tab to another file (and not yet back)

I'm pretty sure there is no GCC extension used. `gnu++11` is just the default from CMake (we should change that). But you are right with the `(Global Scope)`. For class1 it works, for class2 it does not.

I tried replacing the whole file contents of class2.hpp by `class foo{   };` and saved, switched tabs etc. but could not get it to work.

**Comment by bobbrow** [comment]
Headers outside the workspace are not parsed by default.  They will be parsed if you open them in the editor or another file includes them.

Ok, so now that we know there's a tag parser failure involved, you may be able to help us figure out the symbol that is causing the failure.  It's usually related to macros, but it doesn't have to be.  If you start from the top of class2's header file and move the cursor down line by line, you should be able to see the status bar's symbol location update when you are inside a class or function.  The location where it stops updating as expected is the approximate location of the problem.  Unfortunately, we don't turn the DB icon back on when we update the DB for modified files (because it's usually a pretty fast operation), so you might just wait a few seconds to ensure the file has been tag parsed before doing this.

**Comment by Flamefire** [comment]
I have the feeling it is not parsed at all although it is included the the component.cpp. As I said: I replaced the whole contents of that header and it still failed to parse that. I'll try again on monday and wait longer than I did before but I doubt it will change anything. Other ideas? Is there debug output of the tag parser?
BTW: What do you mean with the DB icon?

**Comment by bobbrow** [comment]
You can set `"C_Cpp.loggingLevel": "Debug"`, run the "Reset IntelliSense Database" command and then look in the Output window with the C/C++ filter applied.  You should see a bunch of "tag parsing file:" messages.  Search the output for your header.  As long as it is included by at least one file in the workspace, it should be parsed.

**Comment by Flamefire** [comment]
Note: I had to restart VSC to have the log level have any effect. It seems I also had to explicitly use
```
    "C_Cpp.intelliSenseEngine": "Default",
    "C_Cpp.intelliSenseEngineFallback": "Enabled",
```
However this did not help:
- I had reopened the folder and my symbols (including navigation) worked
- Navigation to include files broken: Not working at all or e.g. `#include <foobar/event.hpp>` jumped into a boost header that had a `event* classmember_` line. Looks like the parser thinks it is a C++ symbol rather than a header
- Apply Debug and above settings and multiple restarts: navigation more or less completely broken. The header files in question don't show up in the log at all (although there are lots of `tag parsing` entries)

**Comment by bobbrow** [comment]
> I had reopened the folder and my symbols (including navigation) worked

So the problem was fixed by reloading the window?

> Edit: I switched to C_Cpp.intelliSenseEngine": "Tag Parser", and GoTo started working.

That's strange... GotoDef shouldn't be affected by that setting.

I'm going to move this to a new issue since the current issue is about something else.

**Comment by sean-mcmanus** [comment]
Go to Defintion on std::vector<>::cend() should work with 0.19.0. I don't think the browse.path population with compile_commands.json was implemented though.

**Comment by elahehrashedi** [comment]
This issue is fixed in 0.28.0.
0.28.0 release: https://github.com/microsoft/vscode-cpptools/releases/tag/0.28.0
